File: sdk/java/src/main/java/com/xxxxteam/tempmail/providers/TempMailPro.java
package com.xxxxteam.tempmail.providers;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.xxxxteam.tempmail.Email;
import com.xxxxteam.tempmail.EmailInfo;
import com.xxxxteam.tempmail.HttpClient;
import com.xxxxteam.tempmail.HttpResult;
import com.xxxxteam.tempmail.Json;
import com.xxxxteam.tempmail.Normalizer;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * TempMailPro 渠道 — https://tempmailpro.us
 * POST /api/v1/mailbox/create 创建邮箱，GET /api/v1/mailbox/{token}/emails 获取邮件。
 */
public final class TempMailPro {

    private static final String API_BASE = "https://tempmailpro.us/api/v1";
    private static final Map<String, String> HEADERS = new LinkedHashMap<>();

    static {
        HEADERS.put("Accept", "application/json");
        HEADERS.put("User-Agent", "Mozilla/5.0");
    }

    private TempMailPro() {
    }

    /**
     * 创建临时邮箱。
     *
     * @return 邮箱信息
     */
    public static EmailInfo generate() {
        Map<String, String> h = new LinkedHashMap<>(HEADERS);
        h.put("Content-Type", "application/json");
        HttpResult resp = HttpClient.post(API_BASE + "/mailbox/create", "{}", "application/json", h);
        resp.ensureSuccess();
        JsonElement data = Json.parse(resp.getBody());
        JsonElement box = data != null && data.isJsonObject() ? data.getAsJsonObject().get("data") : null;
        String email = box != null ? Json.str(box, "address").trim() : "";
        String token = box != null ? Json.str(box, "token").trim() : "";
        if (email.isEmpty() || !email.contains("@") || token.isEmpty()) {
            throw new RuntimeException("tempmailpro: 创建邮箱响应无效");
        }
        return new EmailInfo("tempmailpro", email, token, null, null);
    }

    /**
     * 获取邮件列表。
     *
     * @param token 邮箱令牌
     * @param email 邮箱地址
     * @return 邮件列表
     */
    public static List<Email> getEmails(String token, String email) {
        String mailboxToken = (token != null ? token : "").trim();
        String address = (email != null ? email : "").trim();
        if (mailboxToken.isEmpty()) {
            throw new RuntimeException("tempmailpro: token 为空");
        }
        if (address.isEmpty()) {
            throw new RuntimeException("tempmailpro: 邮箱地址为空");
        }
        HttpResult resp = HttpClient.get(
                API_BASE + "/mailbox/" + ProviderUtil.urlEncode(mailboxToken) + "/emails", HEADERS);
        resp.ensureSuccess();
        JsonElement data = Json.parse(resp.getBody());
        JsonArray rows = Json.arr(data, "data");
        if (rows == null) {
            return new ArrayList<>();
        }
        List<Email> result = new ArrayList<>();
        for (JsonElement item : rows) {
            if (item == null || !item.isJsonObject()) {
                continue;
            }
            String msgId = Json.str(item, "id").trim();
            JsonElement detail = null;
            if (!msgId.isEmpty()) {
                detail = fetchDetail(mailboxToken, msgId);
            }
            Map<String, Object> raw = buildRaw(detail != null ? detail : item, address);
            result.add(Normalizer.normalizeEmail(raw, address));
        }
        return result;
    }

    private static JsonElement fetchDetail(String token, String messageId) {
        try {
            HttpResult resp = HttpClient.get(
                    API_BASE + "/mailbox/" + ProviderUtil.urlEncode(token)
                            + "/emails/" + ProviderUtil.urlEncode(messageId), HEADERS);
            if (!resp.isOk()) {
                return null;
            }
            JsonElement data = Json.parse(resp.getBody());
            if (data != null && data.isJsonObject() && data.getAsJsonObject().has("data")) {
                return data.getAsJsonObject().get("data");
            }
            return null;
        } catch (Exception e) {
            return null;
        }
    }

    private static Map<String, Object> buildRaw(JsonElement el, String recipient) {
        Map<String, Object> raw = Json.toDict(el);
        raw.putIfAbsent("from", raw.getOrDefault("from_address", raw.getOrDefault("from_name", "")));
        raw.putIfAbsent("to", recipient);
        raw.putIfAbsent("text", raw.getOrDefault("body_text", ""));
        raw.putIfAbsent("html", raw.getOrDefault("body_html", ""));
        raw.putIfAbsent("date", raw.getOrDefault("received_at", ""));
        return raw;
    }
}

File: sdk/kotlin/src/main/kotlin/com/xxxxteam/tempmail/providers/TempMailPro.kt
package com.xxxxteam.tempmail.providers

import com.xxxxteam.tempmail.Email
import com.xxxxteam.tempmail.EmailInfo
import com.xxxxteam.tempmail.Normalize
import com.xxxxteam.tempmail.Provider
import kotlinx.serialization.json.*

/**
 * TempMailPro 渠道 — https://tempmailpro.us
 *
 * POST /api/v1/mailbox/create 创建邮箱，GET /api/v1/mailbox/{token}/emails 获取邮件。
 */
object TempMailPro : Provider {

    private const val CHANNEL = "tempmailpro"
    private const val API_BASE = "https://tempmailpro.us/api/v1"

    private val headers = mapOf(
        "Accept" to "application/json",
        "User-Agent" to "Mozilla/5.0",
    )

    /** 创建临时邮箱。 */
    override suspend fun generate(): EmailInfo {
        val h = headers + ("Content-Type" to "application/json")
        val resp = ProviderUtil.httpPost("$API_BASE/mailbox/create", "{}", "application/json", h)
        resp.ensureSuccess()
        val box = ProviderUtil.parseObject(resp.body)?.get("data") as? JsonObject
        val email = box?.let { ProviderUtil.str(it, "address").trim() } ?: ""
        val token = box?.let { ProviderUtil.str(it, "token").trim() } ?: ""
        if (email.isEmpty() || !email.contains("@") || token.isEmpty()) {
            throw RuntimeException("tempmailpro: 创建邮箱响应无效")
        }
        return EmailInfo(email = email, channel = CHANNEL, token = token)
    }

    /** 获取邮件列表。 */
    override suspend fun getEmails(info: EmailInfo): List<Email> {
        val mailboxToken = info.token.trim()
        val address = info.email.trim()
        if (mailboxToken.isEmpty()) throw RuntimeException("tempmailpro: token 为空")
        if (address.isEmpty()) throw RuntimeException("tempmailpro: 邮箱地址为空")
        val resp = ProviderUtil.httpGet("$API_BASE/mailbox/${ProviderUtil.urlEncode(mailboxToken)}/emails", headers)
        resp.ensureSuccess()
        val rows = ProviderUtil.arr(ProviderUtil.parseObject(resp.body), "data") ?: return emptyList()
        val result = mutableListOf<Email>()
        for (item in rows.filterIsInstance<JsonObject>()) {
            val msgId = ProviderUtil.str(item, "id").trim()
            val detail = if (msgId.isNotEmpty()) fetchDetail(mailboxToken, msgId) else null
            result.add(Normalize.fromMap(buildRaw(detail ?: item, address), address))
        }
        return result
    }

    private suspend fun fetchDetail(token: String, messageId: String): JsonObject? {
        return try {
            val resp = ProviderUtil.httpGet(
                "$API_BASE/mailbox/${ProviderUtil.urlEncode(token)}/emails/${ProviderUtil.urlEncode(messageId)}",
                headers,
            )
            if (!resp.isOk) null else ProviderUtil.parseObject(resp.body)?.get("data") as? JsonObject
        } catch (_: Exception) {
            null
        }
    }

    private fun buildRaw(raw: JsonObject, recipient: String): Map<String, Any?> {
        return mapOf(
            "id" to ProviderUtil.str(raw, "id"),
            "from" to ProviderUtil.firstNonEmpty(
                ProviderUtil.str(raw, "from"),
                ProviderUtil.str(raw, "from_address"),
                ProviderUtil.str(raw, "from_name"),
            ),
            "to" to ProviderUtil.firstNonEmpty(ProviderUtil.str(raw, "to"), recipient),
            "subject" to ProviderUtil.str(raw, "subject"),
            "text" to ProviderUtil.firstNonEmpty(ProviderUtil.str(raw, "text"), ProviderUtil.str(raw, "body_text")),
            "html" to ProviderUtil.firstNonEmpty(ProviderUtil.str(raw, "html"), ProviderUtil.str(raw, "body_html")),
            "date" to ProviderUtil.firstNonEmpty(ProviderUtil.str(raw, "date"), ProviderUtil.str(raw, "received_at")),
        )
    }
}
